Restore very old backup fails

https://forum.iobroker.net/topic/54067/wiederherstellung-error-abbruch/2

```
backitup.0
2022-04-09 00:43:59.660 error [ERROR] [iobroker] at BackupRestore._restoreAfterStop (/opt/iobroker/node_modules/iobroker.js-controller/lib/setup/setupBackup.js:686:52)
backitup.0
2022-04-09 00:43:59.658 error [ERROR] [iobroker] TypeError: Cannot read property 'hostname' of undefined
backitup.0
2022-04-09 00:43:59.655 error [ERROR] [iobroker] This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason:
```

--- a/packages/cli/src/lib/setup/setupBackup.ts
+++ b/packages/cli/src/lib/setup/setupBackup.ts
@@ -695,7 +695,7 @@ export class BackupRestore {
         // check that the same controller version is installed as it is contained in backup
         const exitCode = this._ensureCompatibility(
             controllerDir,
-            restore.config?.system ? restore.config.system.hostname || hostname : hostname,
+            restore.config?.system?.hostname || hostname,
             restore.objects,
             force
         );

--- a/packages/cli/src/lib/setup/setupBackup.ts
+++ b/packages/cli/src/lib/setup/setupBackup.ts
@@ -32,7 +32,7 @@ export interface CLIBackupRestoreOptions {
     restartController: () => void;
 }
 
-interface CreateBackupInternalResult {
+interface BackupObject {
     config?: null | Record<string, any>;
     objects: null | ioBroker.GetObjectListItem[];
     states: Record<string, ioBroker.State>;
@@ -258,7 +258,7 @@ export class BackupRestore {
             }
         }
 
-        let result: CreateBackupInternalResult | null = { objects: null, states: {} };
+        let result: BackupObject | null = { objects: null, states: {} };
 
         const hostname = tools.getHostName();
 
@@ -450,7 +450,7 @@ export class BackupRestore {
      * @param statesList - list of state ids
      * @param stateObjects - list of state objects
      */
-    private async _setStateHelper(statesList: string[], stateObjects: ioBroker.State[]): Promise<void> {
+    private async _setStateHelper(statesList: string[], stateObjects: Record<string, ioBroker.State>): Promise<void> {
         for (let i = 0; i < statesList.length; i++) {
             try {
                 // @ts-expect-error #1917 TODO revisit after merge
@@ -679,7 +679,7 @@ export class BackupRestore {
         // replace all hostnames of instances etc with the new host
         data = data.replace(/\$\$__hostname__\$\$/g, hostname);
         fs.writeFileSync(`${tmpDir}/backup/backup_.json`, data);
-        let restore;
+        let restore: BackupObject;
         try {
             restore = JSON.parse(data);
         } catch (err) {
@@ -687,10 +687,15 @@ export class BackupRestore {
             return EXIT_CODES.CANNOT_RESTORE_BACKUP;
         }
 
+        if (!restore.objects) {
+            console.error('The backup does not contain any objects.');
+            return EXIT_CODES.CANNOT_RESTORE_BACKUP;
+        }
+
         // check that the same controller version is installed as it is contained in backup
         const exitCode = this._ensureCompatibility(
             controllerDir,
-            restore.config ? restore.config.system.hostname || hostname : hostname,
+            restore.config?.system ? restore.config.system.hostname || hostname : hostname,
             restore.objects,
             force
         );

